client/helpers_test.go

```diff
@@ -147,3 +148,431 @@ func TestApplyChangelistMulti(t *testing.T) {
+
+	tgts := repo.Targets[data.CanonicalTargetsRole]
+	assert.Len(t, tgts.Signed.Delegations.Roles, 1)
+	assert.Len(t, tgts.Signed.Delegations.Keys, 1)
+
+	_, ok := tgts.Signed.Delegations.Keys[newKey2.ID()]
+	assert.True(t, ok)
+
+	role := tgts.Signed.Delegations.Roles[0]
+	assert.Len(t, role.KeyIDs, 1)
+	assert.Equal(t, newKey2.ID(), role.KeyIDs[0])
+	assert.Equal(t, "targets/level1", role.Name)
+	assert.Equal(t, "level1", role.Paths[0])
+}
+
+func TestApplyTargetsDelegationInvalidRole(t *testing.T) {
+	_, repo, cs := testutils.EmptyRepo()
+
+	newKey, err := cs.Create("targets/level1", data.ED25519Key)
+	assert.NoError(t, err)
+
+	// create delegation
+	kl := data.KeyList{newKey}
+	td := &changelist.TufDelegation{
+		NewThreshold: 1,
+		AddKeys:      kl,
+		AddPaths:     []string{"level1"},
+	}
+
+	tdJSON, err := json.Marshal(td)
+	assert.NoError(t, err)
+
+	ch := changelist.NewTufChange(
+		changelist.ActionCreate,
+		"bad role",
+		changelist.TypeTargetsDelegation,
+		"",
+		tdJSON,
+	)
+
+	err = applyTargetsChange(repo, ch)
+	assert.Error(t, err)
+}
+
+func TestApplyTargetsDelegationInvalidJSONContent(t *testing.T) {
+	_, repo, cs := testutils.EmptyRepo()
+
+	newKey, err := cs.Create("targets/level1", data.ED25519Key)
+	assert.NoError(t, err)
+
+	// create delegation
+	kl := data.KeyList{newKey}
+	td := &changelist.TufDelegation{
+		NewThreshold: 1,
+		AddKeys:      kl,
+		AddPaths:     []string{"level1"},
+	}
+
+	tdJSON, err := json.Marshal(td)
+	assert.NoError(t, err)
+
+	ch := changelist.NewTufChange(
+		changelist.ActionCreate,
+		"targets/level1",
+		changelist.TypeTargetsDelegation,
+		"",
+		tdJSON[1:],
+	)
+
+	err = applyTargetsChange(repo, ch)
+	assert.Error(t, err)
+}
+
+func TestApplyTargetsDelegationInvalidAction(t *testing.T) {
+	_, repo, _ := testutils.EmptyRepo()
+
+	ch := changelist.NewTufChange(
+		"bad action",
+		"targets/level1",
+		changelist.TypeTargetsDelegation,
+		"",
+		nil,
+	)
+
+	err := applyTargetsChange(repo, ch)
+	assert.Error(t, err)
+}
+
+func TestApplyTargetsChangeInvalidType(t *testing.T) {
+	_, repo, _ := testutils.EmptyRepo()
+
+	ch := changelist.NewTufChange(
+		changelist.ActionCreate,
+		"targets/level1",
+		"badType",
+		"",
+		nil,
+	)
+
+	err := applyTargetsChange(repo, ch)
+	assert.Error(t, err)
+}
+
+func TestApplyTargetsDelegationConflictPathsPrefixes(t *testing.T) {
+	_, repo, cs := testutils.EmptyRepo()
+
+	newKey, err := cs.Create("targets/level1", data.ED25519Key)
+	assert.NoError(t, err)
+
+	// create delegation
+	kl := data.KeyList{newKey}
+	td := &changelist.TufDelegation{
+		NewThreshold: 1,
+		AddKeys:      kl,
+		AddPaths:     []string{"level1"},
+	}
+
+	tdJSON, err := json.Marshal(td)
+	assert.NoError(t, err)
+
+	ch := changelist.NewTufChange(
+		changelist.ActionCreate,
+		"targets/level1",
+		changelist.TypeTargetsDelegation,
+		"",
+		tdJSON,
+	)
+
+	err = applyTargetsChange(repo, ch)
+	assert.NoError(t, err)
+
+	// add prefixes and update
+	td.AddPathHashPrefixes = []string{"abc"}
```

Comment:
Could we add a docstring for this test stating that the conflict is that an error is returned if `PathHashPrefixes` is added to a delegation that already has a path prefix, because a role cannot have both?  Similarly for the one below?

It was not immediately obvious without digigng further in the `Role` code that adding a duplicate path is not an error.